### PR TITLE
Create Protection Profile cmdlet and Protection Profile model

### PR DESCRIPTION
**Pull request by NehaRaje**

### CODE REVIEW

File: src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/PSRecoveryServicesClient/PSRecoveryServicesClientHelper.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                    associatedSubscription = true;
+                    break;
+                }
+            }
+
+            if (!associatedSubscription)
+            {
+                throw new InvalidOperationException(
+                    string.Format(
+                    Properties.Resources.SubscriptionIsNotAssociatedWithTheAccount,
```
Comment (suds05) [inline src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/PSRecoveryServicesClient/PSRecoveryServicesClientHelper.cs]: Dont see resx changes for this. Is it already there. Pls make sure you haven't missed adding a file.

File: src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Properties/Resources.resx [hunk 2]
```diff
@@ -203,4 +203,7 @@ ClientRequestId: {3}</value>
   <data name="VaultCredentialGenerationUnSupported" xml:space="preserve">
     <value>Cannot generate vault credentials for this vault. Download it from the Azure Portal.</value>
   </data>
+  <data name="StorageIsNotAssociatedWithTheAccount" xml:space="preserve">
+    <value>Storage {0} is not associated with the account</value>
```
Comment (suds05) [inline src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Properties/Resources.resx]: Nit: Storage account {0} is not associated with current user account

File: src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/PSRecoveryServicesClient/PSRecoveryServicesClientHelper.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+        /// <summary>
+        /// Validates whether the storage belongs to the currently logged account or not.
+        /// </summary>
+        /// <param name="azureStorageAccount">Storage Account details</param>
+        public static void ValidateStorageAccountAssociation(string azureStorageAccount)
+        {
+            bool associatedAccount = false;
+
+            SubscriptionCloudCredentials creds
```
Comment (suds05) [inline src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/PSRecoveryServicesClient/PSRecoveryServicesClientHelper.cs]: Here you are checking whether the storage account belongs to currently selected subscription or not.  (i.e. the current subscription listed by get-azureaccount)

What we ideally need is that we check in all subscriptions listed in get-azureaccount.

Can we do that.

File: src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Service/CreateAzureSiteRecoveryProtectionProfileObject.cs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+        [Parameter(ParameterSetName = ASRParameterSets.EnterpriseToAzure, Mandatory = true, ValueFromPipeline = true)]
+        [ValidateNotNullOrEmpty]
+        public string RecoveryAzureStorageAccount { get; set; }
+
+        /// <summary>
+        /// Gets or sets Replication Frequency of the Protection Profile in seconds.
+        /// </summary>
+        [Parameter(ParameterSetName = ASRParameterSets.Default, Mandatory = true)]
+        [ValidateNotNullOrEmpty]
+        public int ReplicationFrequencySecond { get; set; }
```
Comment (suds05) [inline src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Service/CreateAzureSiteRecoveryProtectionProfileObject.cs]: ReplicationFrequencyInSeconds would be a better name.

File: src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Service/CreateAzureSiteRecoveryProtectionProfileObject.cs [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        [Parameter(ParameterSetName = ASRParameterSets.Default, Mandatory = true)]
+        [ValidateNotNullOrEmpty]
+        public int ReplicationPort { get; set; }
+
+        /// <summary>
+        /// Gets or sets Replication Start time of the Protection Profile.
+        /// </summary>
+        [Parameter(ParameterSetName = ASRParameterSets.Default, Mandatory = true)]
+        [ValidateNotNullOrEmpty]
+        public int ReplicationStartTime { get; set; }
```
Comment (suds05) [inline src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Service/CreateAzureSiteRecoveryProtectionProfileObject.cs]: we generally model time as a DateTime field. Moreover, in Portal, there's a value called 'Immediately'. How do we model that?

I suggest you have two fields:
bool? StartReplicationImmediately
DateTime? ReplicationStartTime.

You should have validations to make sure 
1. If former is true, latter is null, and vice versa.
2. If latter is specified, it is in 1/2 hour intervals like in portal.

I am also open to different models....

File: src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Service/CreateAzureSiteRecoveryProtectionProfileObject.cs [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+            try
+            {
+                switch (this.ReplicationProvider)
+                {
+                    case "HyperVReplica":
+                        this.EnterpriseToEnterpriseProtectionProfileObject();
+                        break;
+                    case "HyperVReplicaAzure":
+                        this.EnterpriseToAzureProtectionProfileObject();
+                        break;
```
Comment (suds05) [inline src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Service/CreateAzureSiteRecoveryProtectionProfileObject.cs]: Add a default case in general and throw an exception that this is an unsupported option.

File: src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Service/CreateAzureSiteRecoveryProtectionProfileObject.cs [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+using Microsoft.WindowsAzure.Management.SiteRecovery.Models;
+using Microsoft.WindowsAzure.Management.Storage.Models;
+
+namespace Microsoft.Azure.Commands.RecoveryServices
+{
+    /// <summary>
+    /// Creates Azure Site Recovery Protection Profile object in memory.
+    /// </summary>
+    [Cmdlet(VerbsCommon.New, "AzureSiteRecoveryProtectionProfile", DefaultParameterSetName = ASRParameterSets.Default)]
+    [OutputType(typeof(ASRProtectionProfile))]
```
Comment (suds05) [inline src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Service/CreateAzureSiteRecoveryProtectionProfileObject.cs]: There need to be different parameter sets for E2E, E2A, SMB2A, etc. I.e one for each provider.

File: src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Service/CreateAzureSiteRecoveryProtectionProfileObject.cs [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+        [Parameter(ParameterSetName = ASRParameterSets.EnterpriseToAzure, Mandatory = true)]
+        [ValidateNotNullOrEmpty]
+        public string RecoveryAzureSubscription { get; set; }
+
+        /// <summary>
+        /// Gets or sets Recovery Azure Storage Account Name of the Protection Profile for E2A scenarios.
+        /// </summary>
+        [Parameter(ParameterSetName = ASRParameterSets.EnterpriseToAzure, Mandatory = true, ValueFromPipeline = true)]
+        [ValidateNotNullOrEmpty]
+        public string RecoveryAzureStorageAccount { get; set; }
```
Comment (suds05) [inline src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Service/CreateAzureSiteRecoveryProtectionProfileObject.cs]: We are mixing up settings of different providers in the same class. E.g. the Storage Account is not valid for SAN, E2E, etc. This is not good - it becomes a mish-mash of different properties.

We should model settings unique to one provider in a sub-object.

I.e. base object has
HyperVReplicaProviderSettings
HyperVReplicaAzureProviderSettings

Put unique objects of each inside them.

File: src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Service/CreateAzureSiteRecoveryProtectionProfileObject.cs [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+    {
+        #region Parameters
+
+        /// <summary>
+        /// Gets or sets Replication Provider of the Protection Profile.
+        /// </summary>
+        [Parameter(ParameterSetName = ASRParameterSets.Default, Mandatory = true)]
+        [ValidateNotNullOrEmpty]
+        [ValidateSet(
+            "HyperVReplica",
```
Comment (sanjkuma) [inline src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Service/CreateAzureSiteRecoveryProtectionProfileObject.cs]: We have these value present in Constants class.

File: src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Service/CreateAzureSiteRecoveryProtectionProfileObject.cs [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+        public string ReplicationProvider { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value for Replication Method of the Protection Profile.
+        /// </summary>
+        [Parameter(ParameterSetName = ASRParameterSets.EnterpriseToEnterprise, Mandatory = true)]
+        [Parameter(ParameterSetName = ASRParameterSets.EnterpriseToAzure, Mandatory = true)]
+        [ValidateNotNullOrEmpty]
+        [ValidateSet(
+            "Online",
```
Comment (sanjkuma) [inline src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Service/CreateAzureSiteRecoveryProtectionProfileObject.cs]: Constants

File: src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/lib/PSObjects.cs [hunk 11]
```diff
@@ -314,52 +330,52 @@ public ASRProtectionContainer(ProtectionContainer pc)
                         var details = DataContractUtils<HyperVReplicaAzureProtectionProfileDetails>.Deserialize(
                             profile.ReplicationProviderSetting);
 
-                        asrProtectionProfile.AllowReplicaDeletion = false;
-                        asrProtectionProfile.ReplicationPort = 0;
+                        asrProtectionProfile.HyperVReplicaAzureProviderSettingsObject.AllowReplicaDeletion = false;
```
Comment (sanjkuma) [inline src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/lib/PSObjects.cs]: New HyperVReplicaAzureProviderSettingsObject()

File: src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/lib/PSObjects.cs [hunk 12]
```diff
@@ -421,6 +437,245 @@ public ASRProtectionContainer(ProtectionContainer pc)
...
+    }
+
+    /// <summary>
+    /// Azure Site Recovery Protection Profile HyperVReplicaProviderSettings.
+    /// </summary>
+    [SuppressMessage(
+        "Microsoft.StyleCop.CSharp.MaintainabilityRules",
+        "SA1402:FileMayOnlyContainASingleClass",
+        Justification = "Keeping all related objects together.")]
+    public class HyperVReplicaProviderSettings
```
Comment (sanjkuma) [inline src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/lib/PSObjects.cs]: I think this object is only used as contract, if so then we can move it to contract file.

**Pull request merged**